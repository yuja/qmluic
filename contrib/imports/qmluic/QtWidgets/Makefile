JQ = jq
QMAKE = qmake

QT_INSTALL_BINS = $(shell $(QMAKE) -query QT_INSTALL_BINS)
QT_INSTALL_LIBS = $(shell $(QMAKE) -query QT_INSTALL_LIBS)
QT_VERSION = $(shell $(QMAKE) -query QT_VERSION)

QMLTYPEREGISTRAR = $(QT_INSTALL_BINS)/qmltyperegistrar
QT_METATYPES = $(QT_INSTALL_LIBS)/metatypes

major_version = $(word 1,$(subst ., ,$(QT_VERSION)))
qtx = qt$(major_version)

.PHONY: all
all: plugins.qmltypes

plugins.qmltypes: metatypes.json $(QT_METATYPES)/$(qtx)core_metatypes.json
	$(QMLTYPEREGISTRAR) \
		--import-name qmluic.QtWidgets \
		--major-version $(major_version) \
		--foreign-types $(QT_METATYPES)/$(qtx)core_metatypes.json \
		--generate-qmltypes $@ $< > /dev/null

metatypes.json: $(QT_METATYPES)/$(qtx)widgets_metatypes.json
	$(JQ) 'map(.classes |= map({"classInfos": [{"name": "QML.Element", "value": "auto"}]} + .))' $< > $@

.PHONY: clean
clean:
	$(RM) plugins.qmltypes metatypes.json
