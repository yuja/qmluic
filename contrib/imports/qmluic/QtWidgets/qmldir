module qmluic.QtWidgets
typeinfo plugins.qmltypes
depends QtQml
