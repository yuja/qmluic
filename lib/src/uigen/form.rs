use super::object::UiObject;
use super::xmlutil;
use super::{FileNameRules, XmlResult, XmlWriter};
use crate::typemap::{Class, TypeSpace};
use quick_xml::events::{BytesStart, Event};
use std::io;

/// Top-level object wrapper to be serialized to UI XML.
#[derive(Clone, Debug)]
pub struct UiForm {
    pub class: Option<String>,
    pub root_object: UiObject,
    pub custom_widgets: Vec<CustomWidget>,
}

impl UiForm {
    /// Serializes this to UI XML.
    pub fn serialize_to_xml<W>(&self, writer: &mut XmlWriter<W>) -> XmlResult<()>
    where
        W: io::Write,
    {
        let tag = BytesStart::borrowed_name(b"ui").with_attributes([("version", "4.0")]);
        writer.write_event(Event::Start(tag.to_borrowed()))?;
        if let Some(name) = &self.class {
            xmlutil::write_tagged_str(writer, "class", name)?;
        }
        self.root_object.serialize_to_xml(writer)?;
        if !self.custom_widgets.is_empty() {
            let tag = BytesStart::borrowed_name(b"customwidgets");
            writer.write_event(Event::Start(tag.to_borrowed()))?;
            for w in &self.custom_widgets {
                w.serialize_to_xml(writer)?;
            }
            writer.write_event(Event::End(tag.to_end()))?;
        }
        writer.write_event(Event::End(tag.to_end()))?;
        writer.write(b"\n")
    }
}

/// User type referenced from the [`UiForm`].
#[derive(Clone, Debug)]
pub struct CustomWidget {
    pub class: String,
    pub extends: String,
    pub header: String,
}

impl CustomWidget {
    pub(super) fn from_class(cls: &Class, file_name_rules: &FileNameRules) -> Option<Self> {
        // If super class doesn't exist, diagnostic message would have been emitted while
        // building the object representation. So returns silently.
        let super_cls = cls.public_super_classes().next()?;
        Some(CustomWidget {
            class: cls.qualified_cxx_name().into(),
            extends: super_cls.qualified_cxx_name().into(),
            header: file_name_rules.type_name_to_cxx_header_name(cls.name()),
        })
    }

    /// Serializes this to UI XML.
    pub fn serialize_to_xml<W>(&self, writer: &mut XmlWriter<W>) -> XmlResult<()>
    where
        W: io::Write,
    {
        let tag = BytesStart::borrowed_name(b"customwidget");
        writer.write_event(Event::Start(tag.to_borrowed()))?;
        xmlutil::write_tagged_str(writer, "class", &self.class)?;
        xmlutil::write_tagged_str(writer, "extends", &self.extends)?;
        xmlutil::write_tagged_str(writer, "header", &self.header)?;
        writer.write_event(Event::End(tag.to_end()))
    }
}
