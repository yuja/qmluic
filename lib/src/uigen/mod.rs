//! Qt user interface XML (.ui) generator.

use crate::diagnostic::{Diagnostic, Diagnostics};
use crate::qmlast::{UiImportSource, UiProgram};
use crate::qmldir;
use crate::qmldoc::UiDocument;
use crate::typemap::{Class, Module, ModuleData, ModuleId, Type, TypeMap, TypeSpace};
use thiserror::Error;

mod expr;
mod gadget;
mod layout;
mod object;
mod property;
mod xmlutil;

pub use self::expr::*; // re-export
pub use self::gadget::*; // re-export
pub use self::layout::*; // re-export
pub use self::object::*; // re-export

pub type XmlError = quick_xml::Error;
pub type XmlResult<T> = quick_xml::Result<T>;
pub type XmlWriter<W> = quick_xml::Writer<W>;

/// Builds `UiForm` from the given `doc`.
pub fn build(
    base_ctx: &BuildContext,
    doc: &UiDocument,
    diagnostics: &mut Diagnostics,
) -> Option<UiForm> {
    let program = diagnostics.consume_err(UiProgram::from_node(doc.root_node(), doc.source()))?;
    // TODO: do not build unnamed module space if directory module available
    let module_data = make_doc_module_data(doc, &program, base_ctx.type_map, diagnostics);
    let type_space = get_doc_type(base_ctx.type_map, doc)
        .unwrap_or_else(|| Type::Module(Module::new(&module_data, base_ctx.type_map)));
    let ctx = BuildDocContext::new(doc, type_space, base_ctx);
    let (obj, cls) =
        object::resolve_object_definition(&ctx, program.root_object_node(), diagnostics)?;
    let root_object =
        UiObject::from_object_definition(&ctx, &cls, &obj, ContainerKind::Any, diagnostics)?;
    Some(UiForm {
        class: doc.type_name().map(|s| s.to_owned()),
        root_object,
    })
}

fn get_doc_type<'a>(type_map: &'a TypeMap, doc: &UiDocument) -> Option<Type<'a>> {
    if let Some(doc_path) = doc.path() {
        let base_dir =
            qmldir::normalize_path(doc_path.parent().expect("doc path should point to file"));
        type_map
            .get_module(ModuleId::Directory(base_dir.into()))
            .and_then(|m| {
                m.get_type(
                    doc.type_name()
                        .expect("doc loaded from file should have type"),
                )
            })
    } else {
        None
    }
}

fn make_doc_module_data(
    doc: &UiDocument,
    program: &UiProgram,
    type_map: &TypeMap,
    diagnostics: &mut Diagnostics,
) -> ModuleData {
    let mut module_data = ModuleData::with_builtins();
    for imp in program.imports() {
        if imp.alias().is_some() {
            diagnostics.push(Diagnostic::error(
                imp.node().byte_range(),
                "aliased import is not supported",
            ));
            continue;
        }
        // TODO: warn that version field is ignored
        // TODO: anchor diagnostic message onto imp.source() node
        let id = match imp.source() {
            UiImportSource::Identifier(x) => {
                ModuleId::Named(x.to_string(doc.source()).into_owned().into())
            }
            UiImportSource::String(x) => {
                if let Some(p) = doc.path().and_then(|p| p.parent()) {
                    ModuleId::Directory(qmldir::normalize_path(p.join(&x)).into())
                } else {
                    diagnostics.push(Diagnostic::error(
                        imp.node().byte_range(),
                        "cannot resolve directory path against inline QML document",
                    ));
                    continue;
                }
            }
        };
        if type_map.contains_module(&id) {
            module_data.import_module(id);
        } else {
            diagnostics.push(Diagnostic::error(
                imp.node().byte_range(),
                "module not found",
            ));
        }
    }
    module_data
}

/// Resources passed around the UI object constructors.
#[derive(Clone, Debug)]
pub struct BuildContext<'a> {
    type_map: &'a TypeMap,
    action_class: Class<'a>,
    form_layout_class: Class<'a>,
    grid_layout_class: Class<'a>,
    hbox_layout_class: Class<'a>,
    layout_class: Class<'a>,
    layout_attached_class: Class<'a>,
    push_button_class: Class<'a>,
    spacer_item_class: Class<'a>,
    tab_widget_class: Class<'a>,
    tab_widget_attached_class: Class<'a>,
    vbox_layout_class: Class<'a>,
    widget_class: Class<'a>,
}

#[derive(Clone, Debug)]
struct BuildDocContext<'a, 's> {
    source: &'s str,
    action_class: Class<'a>,
    form_layout_class: Class<'a>,
    grid_layout_class: Class<'a>,
    hbox_layout_class: Class<'a>,
    layout_class: Class<'a>,
    layout_attached_class: Class<'a>,
    push_button_class: Class<'a>,
    spacer_item_class: Class<'a>,
    tab_widget_class: Class<'a>,
    tab_widget_attached_class: Class<'a>,
    vbox_layout_class: Class<'a>,
    widget_class: Class<'a>,
    type_space: Type<'a>,
}

impl<'a> BuildContext<'a> {
    /// Sets up context with the given `type_map`.
    pub fn prepare(type_map: &'a TypeMap) -> Result<Self, BuildContextError> {
        const MODULE_NAME: &str = "qmluic.QtWidgets";
        let module = type_map
            .get_module(&ModuleId::Named(MODULE_NAME.into()))
            .ok_or(BuildContextError::ModuleNotFound(MODULE_NAME))?;
        let get_class = |name| {
            if let Some(Type::Class(cls)) = module.get_type(name) {
                Ok(cls)
            } else {
                Err(BuildContextError::ClassNotFound(name))
            }
        };
        Ok(BuildContext {
            type_map,
            action_class: get_class("QAction")?,
            form_layout_class: get_class("QFormLayout")?,
            grid_layout_class: get_class("QGridLayout")?,
            hbox_layout_class: get_class("QHBoxLayout")?,
            layout_class: get_class("QLayout")?,
            layout_attached_class: get_class("QLayoutAttached")?,
            push_button_class: get_class("QPushButton")?,
            spacer_item_class: get_class("QSpacerItem")?,
            tab_widget_class: get_class("QTabWidget")?,
            tab_widget_attached_class: get_class("QTabWidgetAttached")?,
            vbox_layout_class: get_class("QVBoxLayout")?,
            widget_class: get_class("QWidget")?,
        })
    }
}

impl<'a, 's> BuildDocContext<'a, 's> {
    fn new(doc: &'s UiDocument, type_space: Type<'a>, base_ctx: &BuildContext<'a>) -> Self {
        BuildDocContext {
            source: doc.source(),
            action_class: base_ctx.action_class.clone(),
            form_layout_class: base_ctx.form_layout_class.clone(),
            grid_layout_class: base_ctx.grid_layout_class.clone(),
            hbox_layout_class: base_ctx.hbox_layout_class.clone(),
            layout_class: base_ctx.layout_class.clone(),
            layout_attached_class: base_ctx.layout_attached_class.clone(),
            push_button_class: base_ctx.push_button_class.clone(),
            spacer_item_class: base_ctx.spacer_item_class.clone(),
            tab_widget_class: base_ctx.tab_widget_class.clone(),
            tab_widget_attached_class: base_ctx.tab_widget_attached_class.clone(),
            vbox_layout_class: base_ctx.vbox_layout_class.clone(),
            widget_class: base_ctx.widget_class.clone(),
            type_space,
        }
    }
}

/// Error occurred while setting up [`BuildContext`].
#[derive(Clone, Debug, Error)]
pub enum BuildContextError {
    #[error("required class not found: {0}")]
    ClassNotFound(&'static str),
    #[error("required module not found: {0}")]
    ModuleNotFound(&'static str),
}
