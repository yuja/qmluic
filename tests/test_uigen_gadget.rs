use qmluic::uigen::DynamicBindingHandling;

pub mod common;

#[test]
fn test_brush() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QGraphicsView {
        backgroundBrush.color: "#123abc"
        backgroundBrush.style: Qt.Dense4Pattern
    }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QGraphicsView" name="graphicsView">
      <property name="backgroundBrush">
       <brush brushstyle="Dense4Pattern">
        <color alpha="255">
         <blue>188</blue>
         <green>58</green>
         <red>18</red>
        </color>
       </brush>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_brush_solid_color() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QGraphicsView {
        backgroundBrush: "#80123abc"
    }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QGraphicsView" name="graphicsView">
      <property name="backgroundBrush">
       <brush brushstyle="SolidPattern">
        <color alpha="128">
         <blue>188</blue>
         <green>58</green>
         <red>18</red>
        </color>
       </brush>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_color() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QColorDialog { currentColor: "#123abc" }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QColorDialog" name="colorDialog">
      <property name="currentColor">
       <color alpha="255">
        <blue>188</blue>
        <green>58</green>
        <red>18</red>
       </color>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_color_alpha() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QColorDialog { currentColor: "#80123abc" }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QColorDialog" name="colorDialog">
      <property name="currentColor">
       <color alpha="128">
        <blue>188</blue>
        <green>58</green>
        <red>18</red>
       </color>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_invalid_color() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QColorDialog { currentColor: "#wtf" }
    "###).unwrap_err(), @r###"
    error: invalid hex color
      ┌─ <unknown>:2:30
      │
    2 │ QColorDialog { currentColor: "#wtf" }
      │                              ^^^^^^
    "###);
}

#[test]
fn test_cursor() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget { cursor: Qt.IBeamCursor }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QWidget" name="widget">
      <property name="cursor">
       <cursorShape>IBeamCursor</cursorShape>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_font() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QLabel {
        font.family: "Monospace"
        font.pointSize: 9
        font.weight: 75
        font.italic: false
        font.bold: true
        font.strikeout: true
        font.styleStrategy: QFont.PreferDefault
        font.kerning: true
    }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QLabel" name="label">
      <property name="font">
       <font>
        <bold>true</bold>
        <family notr="true">Monospace</family>
        <italic>false</italic>
        <kerning>true</kerning>
        <pointsize>9</pointsize>
        <strikeout>true</strikeout>
        <stylestrategy>PreferDefault</stylestrategy>
        <weight>75</weight>
       </font>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_icon_themed() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QToolButton { icon.name: "edit-copy" }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QToolButton" name="toolButton">
      <property name="icon">
       <iconset theme="edit-copy">
       </iconset>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_icon_state_pixmaps() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QToolButton {
        icon.normalOff: "normal-off.png"
        icon.normalOn: "normal-on.png"
    }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QToolButton" name="toolButton">
      <property name="icon">
       <iconset>
        <normaloff>normal-off.png</normaloff>
        <normalon>normal-on.png</normalon>
       </iconset>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_palette_color_group() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
        palette.active { window: "black"; windowText: "white" }
    }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QWidget" name="widget">
      <property name="palette">
       <palette>
        <active>
         <colorrole role="Window">
          <brush brushstyle="SolidPattern">
           <color alpha="255">
            <blue>0</blue>
            <green>0</green>
            <red>0</red>
           </color>
          </brush>
         </colorrole>
         <colorrole role="WindowText">
          <brush brushstyle="SolidPattern">
           <color alpha="255">
            <blue>255</blue>
            <green>255</green>
            <red>255</red>
           </color>
          </brush>
         </colorrole>
        </active>
        <disabled>
        </disabled>
        <inactive>
        </inactive>
       </palette>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_palette_default_role() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
        palette.window: "black"
        palette.disabled.window: "gray"
    }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QWidget" name="widget">
      <property name="palette">
       <palette>
        <active>
         <colorrole role="Window">
          <brush brushstyle="SolidPattern">
           <color alpha="255">
            <blue>0</blue>
            <green>0</green>
            <red>0</red>
           </color>
          </brush>
         </colorrole>
        </active>
        <disabled>
         <colorrole role="Window">
          <brush brushstyle="SolidPattern">
           <color alpha="255">
            <blue>128</blue>
            <green>128</green>
            <red>128</red>
           </color>
          </brush>
         </colorrole>
        </disabled>
        <inactive>
         <colorrole role="Window">
          <brush brushstyle="SolidPattern">
           <color alpha="255">
            <blue>0</blue>
            <green>0</green>
            <red>0</red>
           </color>
          </brush>
         </colorrole>
        </inactive>
       </palette>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_pixmap() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QLabel { pixmap: ":/a.png" }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QLabel" name="label">
      <property name="pixmap">
       <pixmap>:/a.png</pixmap>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_size_policy() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
        sizePolicy.horizontalPolicy: QSizePolicy.Preferred
        sizePolicy.verticalPolicy: QSizePolicy.Expanding
    }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QWidget" name="widget">
      <property name="sizePolicy">
       <sizepolicy hsizetype="Preferred" vsizetype="Expanding">
       </sizepolicy>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_size_policy_with_stretch() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
        sizePolicy.horizontalPolicy: QSizePolicy.Preferred
        sizePolicy.verticalPolicy: QSizePolicy.Expanding
        sizePolicy.horizontalStretch: 2
        sizePolicy.verticalStretch: 1
    }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QWidget" name="widget">
      <property name="sizePolicy">
       <sizepolicy hsizetype="Preferred" vsizetype="Expanding">
        <horstretch>2</horstretch>
        <verstretch>1</verstretch>
       </sizepolicy>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_unpaired_size_policy() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
        sizePolicy.horizontalPolicy: QSizePolicy.Expanding
    }
    "###).unwrap_err(), @r###"
    error: both horizontal and vertical policies must be specified
      ┌─ <unknown>:3:5
      │
    3 │     sizePolicy.horizontalPolicy: QSizePolicy.Expanding
      │     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_stretch_without_size_policy() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
        sizePolicy.horizontalStretch: 2
    }
    "###).unwrap_err(), @r###"
    error: cannot specify stretch without horizontal and vertical policies
      ┌─ <unknown>:3:5
      │
    3 │     sizePolicy.horizontalStretch: 2
      │     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_string_list_tr() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QTextBrowser { searchPaths: [qsTr("a"), qsTr("b")] }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QTextBrowser" name="textBrowser">
      <property name="searchPaths">
       <stringlist>
        <string>a</string>
        <string>b</string>
       </stringlist>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_string_list_notr() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QTextBrowser { searchPaths: ["a", "b"] }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QTextBrowser" name="textBrowser">
      <property name="searchPaths">
       <stringlist notr="true">
        <string>a</string>
        <string>b</string>
       </stringlist>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_string_list_empty() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QTextBrowser { searchPaths: [] }
    "###).unwrap(), @r###"
    <ui version="4.0">
     <class>MyType</class>
     <widget class="QTextBrowser" name="textBrowser">
      <property name="searchPaths">
       <stringlist notr="true">
       </stringlist>
      </property>
     </widget>
    </ui>
    "###);
}

#[test]
fn test_string_list_mixed() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QTextBrowser { searchPaths: [qsTr("a"), "b"] }
    "###).unwrap_err(), @r###"
    error: cannot mix bare and translatable strings
      ┌─ <unknown>:2:29
      │
    2 │ QTextBrowser { searchPaths: [qsTr("a"), "b"] }
      │                             ^^^^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_string_list_constructor() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QTextBrowser { searchPaths: [windowTitle] }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_gadget_property_read_expr() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QLabel {
             text: fontEdit.currentFont.family
             QFontComboBox { id: fontEdit }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_gadget_property_write_expr() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QLabel {
             font: {
                 let f = fontEdit.currentFont;
                 f.pointSize = sizeEdit.value;
                 return f;
             }
             QFontComboBox { id: fontEdit }
             QSpinBox { id: sizeEdit }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_gadget_property_write_to_rvalue() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QFontComboBox {
            onCurrentFontChanged: {
                currentFont.pointSize = 100;
            }
        }
        "###,
    );
    insta::assert_snapshot!(
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap_err(), @r###"
    error: rvalue gadget property is not assignable
      ┌─ <unknown>:4:9
      │
    4 │         currentFont.pointSize = 100;
      │         ^^^^^^^^^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_gadget_property_write_to_rvalue_via_local() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QFontComboBox {
            onCurrentFontChanged: {
                let obj = this;
                obj.currentFont.pointSize = 100;
            }
        }
        "###,
    );
    insta::assert_snapshot!(
        common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap_err(), @r###"
    error: rvalue gadget property is not assignable
      ┌─ <unknown>:5:9
      │
    5 │         obj.currentFont.pointSize = 100;
      │         ^^^^^^^^^^^^^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_gadget_binding() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QLabel {
             font.family: familyEdit.text
             font.pointSize: sizeEdit.value
             QLineEdit { id: familyEdit }
             QSpinBox { id: sizeEdit }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_dynamic_gadget_property_sources() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QLabel {
             font.family: (check.checked ? family1Edit : family2Edit).text
             font.pointSize: (check.checked ? size1Edit : size2Edit).value
             QCheckBox { id: check }
             QLineEdit { id: family1Edit }
             QSpinBox { id: size1Edit }
             QLineEdit { id: family2Edit }
             QSpinBox { id: size2Edit }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}
