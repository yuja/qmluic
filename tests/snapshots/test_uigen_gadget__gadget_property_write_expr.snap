---
source: tests/test_uigen_gadget.rs
expression: ui_support_h
---
#pragma once
#include "ui_mytype.h"

namespace UiSupport {
class MyType
{
public:
    MyType(QLabel *root, Ui::MyType *ui): root_(root), ui_(ui) {}

    void setup()
    {
        this->setupLabelFont();
        this->updateLabelFont();
    }

private:
    enum class BindingIndex : unsigned {
        LabelFont,
    };

    void setupLabelFont()
    {
        QObject::connect(this->ui_->fontEdit, &QFontComboBox::currentFontChanged, this->root_, [this]() { this->updateLabelFont(); });
        QObject::connect(this->ui_->sizeEdit, &QSpinBox::valueChanged, this->root_, [this]() { this->updateLabelFont(); });
    }

    void updateLabelFont()
    {
#ifndef QT_NO_DEBUG
        constexpr unsigned index = static_cast<unsigned>(BindingIndex::LabelFont);
        Q_ASSERT_X(!(this->bindingGuard_[index >> 5] & (1U << (index & 0x1f))), __func__, "binding loop detected");
        this->bindingGuard_[index >> 5] |= (1U << (index & 0x1f));
#endif
        this->root_->setFont(this->evalLabelFont());
#ifndef QT_NO_DEBUG
        this->bindingGuard_[index >> 5] &= ~(1U << (index & 0x1f));
#endif
    }

    QFont evalLabelFont()
    {
        QFont a0;
        QFont a1;
        int a2;
    b0:
        a0 = this->ui_->fontEdit->currentFont();
        a1 = a0;
        a2 = this->ui_->sizeEdit->value();
        a1.setPointSize(a2);
        return a1;
    b1:
        Q_UNREACHABLE();
    }

    struct PropertyObserver
    {
        QMetaObject::Connection connection;
        QObject *object = nullptr;
    };

    QLabel *const root_;
    Ui::MyType *const ui_;
#ifndef QT_NO_DEBUG
    quint32 bindingGuard_[1] = {0};
#endif

};
} // namespace UiSupport

